debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/TransactionalBuffer.java

```diff
@@ -70,34 +79,55 @@
     }
 
     /**
-     * Registers callback to execute when transaction commits.
+     * Register a DML operation with the transaction buffer.
      *
-     * @param transactionId transaction identifier
-     * @param scn           SCN
-     * @param changeTime    time of DML parsing completion
-     * @param callback      callback to execute when transaction commits
+     * @param operation operation type
+     * @param transactionId unique transaction identifier
+     * @param scn system change number
+     * @param tableId table identifier
+     * @param parseEntry parser entry
+     * @param changeTime time the DML operation occurred
+     * @param rowId unique row identifier
      */
-    void registerCommitCallback(String transactionId, Scn scn, Instant changeTime, CommitCallback callback) {
+    void registerDmlOperation(int operation, String transactionId, Scn scn, TableId tableId, LogMinerDmlEntry parseEntry, Instant changeTime, String rowId) {
         if (abandonedTransactionIds.contains(transactionId)) {
-            LogMinerHelper.logWarn(streamingMetrics, "Captured DML for abandoned transaction {}, ignored", transactionId);
+            LogMinerHelper.logWarn(streamingMetrics, "Captured DML for abandoned transaction {}, ignored.", transactionId);
             return;
         }
-        // this should never happen
         if (rolledBackTransactionIds.contains(transactionId)) {
-            LogMinerHelper.logWarn(streamingMetrics, "Captured DML for rolled-back transaction {}, ignored", transactionId);
+            LogMinerHelper.logWarn(streamingMetrics, "Captured DML for rolled back transaction {}, ignored.", transactionId);
             return;
         }
 
-        transactions.computeIfAbsent(transactionId, s -> new Transaction(scn));
+        Transaction transaction = transactions.computeIfAbsent(transactionId, s -> new Transaction(transactionId, scn));
+        transaction.events.add(new DmlEvent(operation, parseEntry, scn, tableId, rowId));
 
         streamingMetrics.setActiveTransactions(transactions.size());
         streamingMetrics.incrementRegisteredDmlCount();
         streamingMetrics.calculateLagMetrics(changeTime);
+    }
 
+    /**
+     * Undo a staged DML operation in the transaction buffer.
+     *
+     * @param transactionId unique transaction identifier
+     * @param undoRowId unique row identifier to be undone
+     * @param tableId table identifier
+     */
+    void undoDmlOperation(String transactionId, String undoRowId, TableId tableId) {
```

Comment:
Can't we simply skip records with rollback flag = 1 in the first place? Instead of even adding them to the buffer to begin with?

Reply:
After looking back over the tests in this PR, unfortunately we can't.  I believe I made a small typo earlier in the thread where I gave the example for this constraint violation.  The example should have been
```
TRX SCN   OPERATION ROLLBACK SQL
---- ----- --------- -------- --------------------------------------------------------------------
123 12345 INSERT    0        insert into "DEBEZIUM"."TABLE" ("ID","DATA") values ('1', 'Test');
123 12346 INSERT    0        insert into "DEBEZIUM"."TABLE" ("ID","DATA") values ('1', 'Test');
123 12347 INSERT    1        insert into "DEBEZIUM"."TABLE" ("ID","DATA") values ('1', 'Test');
123 12348 COMMIT
```
If the row for SCN 12346 had actually used rollback flag `= 1` we could have omitted these entirely; however due to the nature of how the rows are materialized for this constraint violation use case, we unfortunately have to filter these inside the buffer after-the-fact.